js/packages/teams-ai/src/authentication/BotAuthentication.ts

```diff
@@ -0,0 +1,280 @@
+            context.activity.text != null &&
+            context.activity.text != undefined &&
+            context.activity.text.length > 0
+        );
+    }
+
+    /**
+     * The handler function is called when the user has successfully signed in
+     * @template TState
+     * @param {(context: TurnContext, state: TState) => Promise<void>} handler The handler function to call when the user has successfully signed in
+     */
+    public onUserSignInSuccess(handler: (context: TurnContext, state: TState) => Promise<void>): void {
+        this._userSignInSuccessHandler = handler;
+    }
+
+    /**
+     * The handler function is called when the user sign in flow fails
+     * @template TState
+     * @param {(context: TurnContext, state: TState) => Promise<void>} handler The handler function to call when the user failed to signed in
+     */
+    public onUserSignInFailure(
+        handler: (context: TurnContext, state: TState, error: AuthError) => Promise<void>
+    ): void {
+        this._userSignInFailureHandler = handler;
+    }
+
+    public async handleSignInActivity(context: TurnContext, state: TState): Promise<void> {
+        try {
+            const userDialogStatePropertyName = this.getUserDialogStatePropertyName(context);
+            const result = await this.runDialog(context, state, userDialogStatePropertyName);
+
+            if (result.status === DialogTurnStatus.complete) {
+                // OAuthPrompt dialog should have sent an invoke response already.
+
+                if (result.result?.token) {
+                    // Successful sign in
+                    setTokenInState(state, this._settingName, result.result.token);
+
+                    // Get user auth state
+                    const userAuthState = state.conversation.value[
+                        this.getUserAuthStatePropertyName(context)
+                    ] as UserAuthState;
+
+                    // Restore previous user message
+                    context.activity.text = userAuthState.message || '';
+
+                    await this._userSignInSuccessHandler?.(context, state);
+                } else {
+                    // Failed sign in
+                    await this._userSignInFailureHandler?.(
+                        context,
+                        state,
+                        new AuthError('Authentication flow completed without a token.', 'completionWithoutToken')
+                    );
+                }
+            }
+        } catch (e) {
+            const errorMessage = e instanceof Error ? e.message : JSON.stringify(e);
+            const message = `Unexpected error encountered while signing in: ${errorMessage}. 
+                Incomming activity details: type: ${context.activity.type}, name: ${context.activity.name}`;
+
+            await this._userSignInFailureHandler?.(context, state, new AuthError(message));
+        }
+    }
+
+    public async runDialog(
```

Comment:
nit: rename to `runAuthDialog`?

Reply:
feel free to resolve if no

Reply:
it's in the BotAuthentication class, so the context should be clear.